cap/healthcheck_test.go

```diff
@@ -0,0 +1,217 @@
+	b1 := cap.NewSupervisorSpec(b1n, cap.WithNodes(cs[2], cs[3]))
+
+	events, err := ObserveSupervisorWithNotifiers(
+		context.TODO(),
+		parentName,
+		cap.WithNodes(
+			cap.Subtree(b0),
+			cap.Subtree(b1),
+		),
+		[]cap.Opt{},
+		[]cap.EventNotifier{func(ev cap.Event) { healthcheckMonitor.HandleEvent(ev) }},
+		func(em EventManager) {},
+	)
+
+	assert.Error(t, err)
+
+	AssertExactMatch(t, events,
+		[]EventP{
+			// start children from left to right
+			WorkerStarted("root/branch0/child0"),
+			WorkerStarted("root/branch0/child1"),
+			SupervisorStarted("root/branch0"),
+			WorkerStarted("root/branch1/child2"),
+			WorkerStarted("root/branch1/child3"),
+			SupervisorStarted("root/branch1"),
+			SupervisorStarted("root"),
+			// NOTE: From here, the stop of the supervisor begins
+			WorkerTerminated("root/branch1/child3"),
+			// NOTE: the child2 never stops and fails with a timeout
+			WorkerFailed("root/branch1/child2"),
+			// NOTE: The supervisor branch1 fails because of child2 timeout
+			SupervisorFailed("root/branch1"),
+			WorkerTerminated("root/branch0/child1"),
+			WorkerTerminated("root/branch0/child0"),
+			SupervisorTerminated("root/branch0"),
+			SupervisorFailed("root"),
+		},
+	)
+
+	// Not healthy
+	assert.False(t, healthcheckMonitor.IsHealthy())
+
+	hr := healthcheckMonitor.GetHealthReport()
+	failedProcesses := hr.GetFailedProcesses()
+	delayedRestartProcesses := hr.GetDelayedRestartProcesses()
+	// Failures are over tolerance. Failed supervisor, branch and child
+	assert.EqualValues(t, 3, len(failedProcesses))
+	assert.True(t, failedProcesses["root/branch1/child2"])
+	assert.True(t, failedProcesses["root/branch1"])
+	assert.True(t, failedProcesses["root"])
+	// restart delays are under tolerance. Nobody restarted
+	assert.EqualValues(t, 3, len(delayedRestartProcesses))
+	assert.True(t, delayedRestartProcesses["root/branch1/child2"])
+	assert.True(t, delayedRestartProcesses["root/branch1"])
+	assert.True(t, delayedRestartProcesses["root"])
+}
+
+func TestHealthPermanentOneForOneNestedFailingWorkerRecovers(t *testing.T) {
+	healthcheckMonitor := cap.NewHealthcheckMonitor(0, 0*time.Millisecond)
+
+	parentName := "root"
+	// Fail only one time
+	child1, failWorker1 := FailOnSignalWorker(1, "child1", cap.WithRestart(cap.Permanent))
+	tree1 := cap.NewSupervisorSpec("subtree1", cap.WithNodes(child1))
+
+	events, err := ObserveSupervisorWithNotifiers(
+		context.TODO(),
+		parentName,
+		cap.WithNodes(cap.Subtree(tree1)),
+		[]cap.Opt{},
+		[]cap.EventNotifier{func(ev cap.Event) { healthcheckMonitor.HandleEvent(ev) }},
+		func(em EventManager) {
+			// NOTE: we won't stop the supervisor until the child has failed at least
+			// once
+			evIt := em.Iterator()
+			// 1) Wait till all the tree is up
+			evIt.SkipTill(SupervisorStarted("root"))
+			// 2) Start the failing behavior of child1
+			failWorker1(true /* done */)
+			// 3) Wait till first restart
+			evIt.SkipTill(WorkerStarted("root/subtree1/child1"))
```

Comment:
I would add a `IsHealthy` call here that returns false, and then returns true after the restart